// react-expand-collapse/dist/ExpandCollapse.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  // the height of data preview
  previewHeight: _propTypes2.default.string.isRequired,
  // any children you pass in
  children: _propTypes2.default.node.isRequired,
  // a custom class name
  className: _propTypes2.default.string,
  // a flag to set default state of data
  expanded: _propTypes2.default.bool,
  // the text to show for expand
  expandText: _propTypes2.default.string,
  // the text to show for collapse
  collapseText: _propTypes2.default.string,
  // option to set if expanded data can be collapsed
  collapse: _propTypes2.default.bool,
  // a flag to set ellipsis
  ellipsis: _propTypes2.default.bool,
  // custom text to show before expand text
  ellipsisText: _propTypes2.default.string,
  // called whenever the user clicks expand button
  onExpandClick: _propTypes2.default.func,
  // called whenever the user clicks collapse button
  onCollapseClick: _propTypes2.default.func,
  // called whenever the user clicks expand and collapse button
  onClick: _propTypes2.default.func
};

var defaultProps = {
  className: "",
  expanded: false,
  expandText: "Expand",
  collapseText: "Collapse",
  collapse: true,
  ellipsis: true,
  ellipsisText: "...",
  onExpandClick: null,
  onCollapseClick: null,
  onClick: null
};

var ExpandCollapse = function (_React$Component) {
  _inherits(ExpandCollapse, _React$Component);

  function ExpandCollapse(props) {
    _classCallCheck(this, ExpandCollapse);

    var _this = _possibleConstructorReturn(this, (ExpandCollapse.__proto__ || Object.getPrototypeOf(ExpandCollapse)).call(this, props));

    _this.state = {
      expanded: _this.props.expanded,
      shouldExpand: true
    };

    _this.handleClick = _this.handleClick.bind(_this);
    _this.shouldDataExpand = _this.shouldDataExpand.bind(_this);
    _this.getContentHeight = _this.getContentHeight.bind(_this);
    _this.getButton = _this.getButton.bind(_this);
    _this.getButtonText = _this.getButtonText.bind(_this);
    _this.getClassName = _this.getClassName.bind(_this);
    _this.setScrollPosition = _this.setScrollPosition.bind(_this);
    _this.setRef = _this.setRef.bind(_this);
    return _this;
  }

  _createClass(ExpandCollapse, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.shouldDataExpand();
    }
  }, {
    key: "handleClick",
    value: function handleClick(event) {
      this.setState({
        expanded: !this.state.expanded
      }, function () {
        var expanded = this.state.expanded;
        var _props = this.props,
            onExpandClick = _props.onExpandClick,
            onCollapseClick = _props.onCollapseClick,
            onClick = _props.onClick;


        if (onExpandClick) {
          if (expanded) {
            onExpandClick(event);
          }
        }

        if (onCollapseClick) {
          if (!expanded) {
            onCollapseClick(event);
          }
        }

        if (onClick) {
          onClick(event);
        }

        this.setScrollPosition();
      });
    }
  }, {
    key: "setScrollPosition",
    value: function setScrollPosition() {
      var expanded = this.state.expanded;


      if (!expanded) {
        var contentRect = this.toggleContent.getBoundingClientRect();
        if (contentRect.top < 0) {
          var offsetTop = Math.abs(contentRect.top + (window.pageYOffset || document.documentElement.scrollTop));

          window.scrollTo(0, offsetTop);
        }
      }
    }
  }, {
    key: "shouldDataExpand",
    value: function shouldDataExpand() {
      var expanded = this.state.expanded;


      var contentRect = this.toggleContent.getBoundingClientRect();
      var contentBodyRect = this.toggleContent.querySelector(".react-expand-collapse__body").getBoundingClientRect();

      if (contentRect.height > contentBodyRect.height && !expanded) {
        this.setState({
          shouldExpand: false
        });
      }
    }
  }, {
    key: "getContentHeight",
    value: function getContentHeight() {
      var _state = this.state,
          expanded = _state.expanded,
          shouldExpand = _state.shouldExpand;
      var previewHeight = this.props.previewHeight;


      if (expanded || !shouldExpand) {
        return "auto";
      }

      return previewHeight;
    }
  }, {
    key: "getButton",
    value: function getButton() {
      var _state2 = this.state,
          expanded = _state2.expanded,
          shouldExpand = _state2.shouldExpand;
      var collapse = this.props.collapse;


      if (shouldExpand) {
        if (!collapse && expanded) {
          return "";
        }

        var buttonText = this.getButtonText();

        return _react2.default.createElement(
          "span",
          {
            className: "react-expand-collapse__button",
            onClick: this.handleClick,
            "aria-label": buttonText,
            "aria-expanded": expanded,
            role: "button"
          },
          buttonText
        );
      }

      return "";
    }
  }, {
    key: "getButtonText",
    value: function getButtonText() {
      var expanded = this.state.expanded;
      var _props2 = this.props,
          expandText = _props2.expandText,
          collapseText = _props2.collapseText,
          ellipsis = _props2.ellipsis,
          ellipsisText = _props2.ellipsisText;


      var text = expanded ? collapseText : expandText;

      if (ellipsis) {
        text = !expanded ? "" + (ellipsisText + " " + text) : text;
      }

      return text;
    }
  }, {
    key: "getClassName",
    value: function getClassName() {
      var expanded = this.state.expanded;
      var className = this.props.className;


      var expandedClass = expanded ? "react-expand-collapse--expanded" : "";

      var classes = ["react-expand-collapse__content", expandedClass].concat(className.split(" ")).join(" ");

      return classes;
    }
  }, {
    key: "setRef",
    value: function setRef(ref) {
      this.toggleContent = ref;
    }
  }, {
    key: "render",
    value: function render() {
      var className = this.getClassName();
      var contentHeight = this.getContentHeight();
      var button = this.getButton();

      return _react2.default.createElement(
        "div",
        {
          className: className,
          ref: this.setRef,
          style: { height: contentHeight }
        },
        _react2.default.createElement(
          "div",
          { className: "react-expand-collapse__body" },
          this.props.children
        ),
        button
      );
    }
  }]);

  return ExpandCollapse;
}(_react2.default.Component);

ExpandCollapse.propTypes = propTypes;
ExpandCollapse.defaultProps = defaultProps;

exports.default = ExpandCollapse;
